### PR TITLE
feat: prevent adhoc metrics from guest user

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

Ensure that guest users (from embedded dashboards) can only use predefined metrics in chart queries.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

Added unit tests covering both `EMBEDDED_SUPERSET` enabled and disabled.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by michael-s-molina** [comment]
Should this be a `fix` instead of `feat`?

File: superset/common/query_object.py [hunk 1]
```diff
@@ -189,6 +191,19 @@ def _set_metrics(self, metrics: list[Metric] | None = None) -> None:
...
 
+        # When connecting with a guest token users should only be allowed to use
+        # predefined metrics.
+        if security_manager.is_guest_user() and any(
+            is_adhoc_metric(metric) for metric in metrics or []
+        ):
+            raise SupersetSecurityException(
+                SupersetError(
+                    error_type=SupersetErrorType.DASHBOARD_SECURITY_ACCESS_ERROR,
+                    message="Guest users can only use predefined metrics",
```
Comment (michael-s-molina) [inline superset/common/query_object.py:202]:
```suggestion
                    message=_("Guest users can only use predefined metrics"),
```

File: tests/unit_tests/queries/query_object_test.py [hunk 2]
```diff
@@ -343,3 +347,68 @@ def test_cache_key_cache_impersonation_on_with_different_user_and_db_impersonati
             ),
         ]
     )
+
+
+def test_query_object_metrics():
```
Comment (michael-s-molina) [inline tests/unit_tests/queries/query_object_test.py:352]: Nice!

**Comment by mistercrunch** [comment]
Does this mean that if I create a dashboard today with a chart in it that uses a SQL expression metric, embed it and share it with guests it would fail? If that's the case it's likely a breaking change. 

Or does this apply only in some other context, say some sort of "explore-embedded" experience?

**Comment by betodealmeida** [comment]
> Does this mean that if I create a dashboard today with a chart in it that uses a SQL expression metric, embed it and share it with guests it would fail? If that's the case it's likely a breaking change.

Yeah, this is a breaking change for security reasons. We need to think of a better way of preventing guest users from modifying the SQL.

**Comment by mistercrunch** [comment]
Doing some thinking here, the user should be able to run what the dashboard designer built for them, in other words, the query that's served on the server side should be ok to run. What we don't want is for the user to be able to alter the chart definition and make up new metrics.

Even with this predefined metrics restriction, the user could in theory point to another pre-defined metric that's out of the scope of the dashboard.

Would an approach of sending a unique identifier of the chart, along with dashboard level filter would work as a better approach here. Meaning the form data sent by the client is simply a reference to a chart + the context from the dashboard, and we can trust that the query executed is the one that the dashboard designer and dashboard "embedder" cleared?

**Comment by betodealmeida** [comment]
Closing in favor of a different approach, see https://github.com/apache/superset/pull/26749.

**Pull request closed without merging**